tests/unit/json_test.c

```diff
@@ -1908,116 +1928,107 @@ static void test_json_null_not_null(void)
-    default:
-        return JsonCompare(a, b) == 0;
-    }
-}
-
 static bool check_json_object_merge_deep(const char *base_raw, const char *extra_raw, const char *expected_raw)
 {
+    LogSetGlobalLevel(LOG_LEVEL_DEBUG);
+    bool pass = true;
+    JsonElement *actual = NULL;
     JsonElement *base = NULL;
+    JsonElement *extra = NULL;
+    JsonElement *expected = NULL;
+    char *expected_string = NULL;
+    char *actual_string = NULL;
+
     if (JsonParse(&base_raw, &base) != JSON_PARSE_OK)
     {
-        return false;
+        printf("error: could not parse base: %s\n", base_raw);
+        pass = false;
+        goto finish;
     }
 
-    JsonElement *extra = NULL;
     if (JsonParse(&extra_raw, &extra) != JSON_PARSE_OK)
     {
-        return false;
+        printf("error: could not parse extra: %s\n", extra_raw);
+        pass = false;
+        goto finish;
     }
 
-    JsonElement *expected = NULL;
     if (JsonParse(&expected_raw, &expected) != JSON_PARSE_OK) {
-        return false;
+        printf("error: could not parse expected: %s\n", expected_raw);
+        pass = false;
+        goto finish;
     }
 
-    JsonElement *actual = JsonObjectMergeDeep(base, extra);
+    expected_string = JsonToString(expected);
+
+    actual = JsonObjectMergeDeep(base, extra);
     if (actual == NULL || actual == base)
     {
-        return false;
+        pass = false;
+        goto finish;
     }
 
-    /* JsonCompare is naive and does not handle containers properly, thus we
-     * need a custom comparison function to test for equality. */
-    if (!compare_json_object_merge_deep(actual, expected))
+    actual_string = JsonToString(actual);
+
+    if (JsonCompare(actual, expected) != 0)
     {
-        return false;
+        pass = false;
+        goto finish;
     }
 
     JsonDestroy(actual);
+    actual = NULL;
     actual = JsonObjectMergeDeepInplace(base, extra);
+
+    // JsonObjectMergeDeepInplace() returns the first parameter (base) so actual should point to base
     if (actual != base)
     {
-        return false;
+        pass = false;
+        goto finish;
     }
 
-    if (!compare_json_object_merge_deep(actual, expected))
+    if (JsonCompare(actual, expected))
     {
-        return false;
+        pass = false;
+        goto finish;
     }
 
-    JsonDestroy(base);
-    JsonDestroy(extra);
-    JsonDestroy(expected);
+    finish:
+        if (actual_string)
+        {
+            if (pass == false)
+            {
+                printf("merged json  : %s\n", actual_string);
+            }
+            free(actual_string);
+        }
+        if (expected_string)
+        {
+            if (pass == false)
+            {
+                printf("expected json: %s\n", expected_string);
+            }
+            free(expected_string);
+        }
+        // in second step of JsonObjectMergeDeepInplace() actual should point to base
+        if (actual != base && actual != NULL)
+        {
+            JsonDestroy(actual);
```

Comment:
`JsonDestroy()` is safe to call with `NULL` so I believe this call and the ones below can be unconditional. **But it's still important to make sure it's not called twice on the same pointer.** (which `!= NULL` doesn't guarantee, of course)

Reply:
I was trying to be careful above since we re-use the actual pointer I set it to NULL so that this later code "knows" that it has been destroyed. Let me know what you think of this strategy or suggestions on a different way to go. I think I initially added these checks to avoid a double-free and why I set to NULL earlier.

line 1980: (cool line number ;)
```
    JsonDestroy(actual);
    actual = NULL;
    actual = JsonObjectMergeDeepInplace(base, extra);
```

Reply:
@craigcomstock we have the macro `DESTROY_AND_NULL` from `alloc.h` for this;

```
DESTROY_AND_NULL(JsonDestroy, actual);
```